File: app/portal.py
"""HTTP client for the Urja Meter Ops portal.

Everything the portal-specific knowledge lives here: login, session cookie,
the HMAC-signed bulk export, rate-limit backoff. See PROTOCOL.md.
"""
import hashlib
import hmac
import os
import threading
import time

import httpx


class PortalError(Exception):
    def __init__(self, status: int, message: str):
        super().__init__(message)
        self.status = status


class NotFound(PortalError):
    pass


class Portal:
    def __init__(self) -> None:
        self.base = os.environ.get("URJA_BASE_URL", "https://urja-ops.flockenergy.tech")
        self.email = os.environ.get("URJA_EMAIL", "operator@urja.local")
        self.password = os.environ.get("URJA_PASSWORD", "urja-ops-2026")
        self.http = httpx.Client(base_url=self.base, timeout=20, follow_redirects=False)
        self._lock = threading.Lock()
        self._signing_secret: str | None = None

    # -- auth ---------------------------------------------------------------
    def login(self) -> None:
        with self._lock:
            self.http.cookies.clear()
            r = self.http.post(
                "/login",
                data={"email": self.email, "password": self.password},
                headers={"Origin": self.base},
            )
            if r.status_code != 200 or "session_token" not in str(r.headers.get("set-cookie", "")):
                raise PortalError(502, f"portal login failed ({r.status_code})")
            self._signing_secret = None

    def _signing_secret_value(self) -> str:
        if not self._signing_secret:
            self._signing_secret = self.get("/portal/keys")["data"]["signingSecret"]
        return self._signing_secret

    # -- core request with re-login, retry and backoff ----------------------
    def get(self, path: str, params: dict | None = None, headers: dict | None = None) -> dict:
        last: httpx.Response | None = None
        for attempt in range(5):
            if not self.http.cookies:
                self.login()
            try:
                r = self.http.get(path, params=params, headers=headers)
            except httpx.HTTPError as e:
                last = None
                time.sleep(0.5 * (attempt + 1))
                if attempt == 4:
                    raise PortalError(502, f"portal unreachable: {e}")
                continue
            last = r
            # expired session: page routes redirect to /login, api routes may 401
            if r.status_code in (302, 401) and "signature" not in r.text:
                self.login()
                continue
            if r.status_code == 429:
                # portal allows roughly 120 requests per 20s window, then ~40s lockout
                time.sleep(min(45, 5 * 2**attempt))
                continue
            if r.status_code == 404:
                raise NotFound(404, _msg(r))
            if r.status_code >= 500:
                time.sleep(0.5 * (attempt + 1))
                continue
            if r.status_code != 200:
                raise PortalError(502, f"portal returned {r.status_code}: {_msg(r)}")
            try:
                return r.json()
            except ValueError:
                raise PortalError(502, "portal returned non-JSON body")
        raise PortalError(503, f"portal kept failing (last status {last.status_code if last else 'n/a'})")

    # -- endpoints ----------------------------------------------------------
    def search_meters(self, q: str = "", page: int = 1) -> dict:
        return self.get("/portal/meters/search", {"q": q, "page": page})

    def geo(self, meter_id: str) -> dict:
        return self.get(f"/portal/meters/{meter_id}/geo")["data"]

    def energy(self, meter_id: str, date_from: str | None, date_to: str | None) -> list[dict]:
        params = {k: v for k, v in {"from": date_from, "to": date_to}.items() if v}
        return self.get(f"/portal/meters/{meter_id}/energy", params)["data"]

    def transformers(self) -> list[dict]:
        out, page = [], 1
        while True:
            r = self.get("/portal/dts", {"page": page})
            out += r["data"]
            if len(out) >= r["total"] or not r["data"]:
                return out
            page += 1

    def export_all(self) -> list[dict]:
        """Bulk export of every meter with hierarchy and geo, HMAC-signed."""
        for attempt in range(2):
            qs = "page=1"
            ts = str(int(time.time()))
            msg = "\n".join(["GET", "/portal/export", qs, ts]).encode()
            sig = hmac.new(self._signing_secret_value().encode(), msg, hashlib.sha256).hexdigest()
            try:
                return self.get("/portal/export?" + qs, headers={"x-timestamp": ts, "x-signature": sig})["data"]
            except PortalError as e:
                # secret may have rotated; refetch once
                if "signature" in str(e) and attempt == 0:
                    self._signing_secret = None
                    continue
                raise
        raise PortalError(502, "export signature rejected")


def _msg(r: httpx.Response) -> str:
    try:
        return r.json().get("message") or r.text[:200]
    except ValueError:
        return r.text[:200]

File: src/portal.ts
// Client for the Urja Meter Ops portal. Everything portal-specific lives here:
// login, the session cookie, the HMAC-signed bulk export, rate-limit backoff.
// See PROTOCOL.md for how each of these was discovered.
import { createHmac } from "node:crypto";
import { setTimeout as sleep } from "node:timers/promises";
import { z } from "zod";

const BASE = process.env.URJA_BASE_URL ?? "https://urja-ops.flockenergy.tech";
const EMAIL = process.env.URJA_EMAIL ?? "operator@urja.local";
const PASSWORD = process.env.URJA_PASSWORD ?? "urja-ops-2026";

export class PortalError extends Error {
  constructor(readonly status: number, message: string) {
    super(message);
  }
}

// Shapes of the portal's responses, validated once at the boundary.
const Node = z.object({ name: z.string(), code: z.string() });
const ExportMeter = z.object({
  meterId: z.string(),
  serialNo: z.string(),
  make: z.string(),
  phaseType: z.string(),
  installStatus: z.string(),
  installType: z.string(),
  build: z.string(),
  dtCode: z.string(),
  hierarchy: z.object({
    zone: Node, circle: Node, division: Node, subdivision: Node,
    substation: Node, feeder: Node, dt: Node,
  }),
  geo: z.object({ lat: z.number(), lng: z.number() }).nullable(),
});
const Transformer = z.object({
  code: z.string(), name: z.string(), feederCode: z.string(), capacityKva: z.number(),
});
const RawReading = z.object({
  timestamp: z.string(), kwh: z.string(), kvah: z.string(), voltR: z.string(),
});

export type ExportMeter = z.infer<typeof ExportMeter>;
export type Transformer = z.infer<typeof Transformer>;
export type RawReading = z.infer<typeof RawReading>;

let cookie = "";
let signingSecret = "";

async function login(): Promise<void> {
  const res = await fetch(`${BASE}/login`, {
    method: "POST",
    redirect: "manual",
    // SvelteKit form actions reject cross-origin posts, so send Origin.
    headers: { "content-type": "application/x-www-form-urlencoded", origin: BASE },
    body: new URLSearchParams({ email: EMAIL, password: PASSWORD }),
  });
  const session = res.headers.getSetCookie().find((c) => c.includes("session_token"));
  if (!session) throw new PortalError(502, `portal login failed (${res.status})`);
  cookie = session.split(";")[0] ?? "";
  signingSecret = "";
}

// GET with re-login on expired session, backoff on 429, retry on 5xx/network.
async function get<T>(path: string, schema: z.ZodType<T>, headers: Record<string, string> = {}): Promise<T> {
  let lastStatus = 0;
  for (let attempt = 0; attempt < 5; attempt++) {
    if (!cookie) await login();
    let res: Response;
    try {
      res = await fetch(BASE + path, {
        headers: { cookie, ...headers },
        redirect: "manual",
        signal: AbortSignal.timeout(20_000),
      });
    } catch {
      await sleep(500 * (attempt + 1));
      continue;
    }
    lastStatus = res.status;
    if (res.status === 200) {
      const parsed = schema.safeParse(await res.json().catch(() => null));
      if (!parsed.success) throw new PortalError(502, `unexpected portal response for ${path}`);
      return parsed.data;
    }
    if (res.status === 404) throw new PortalError(404, "Meter not found");
    if (res.status === 302 || (res.status === 401 && !path.startsWith("/portal/export"))) {
      cookie = ""; // session expired (cookie lives one hour)
    } else if (res.status === 401) {
      throw new PortalError(401, "export signature rejected");
    } else if (res.status === 429) {
      // ~120 requests per 20s, then a ~40s lockout, and no Retry-After header
      await sleep(Math.min(45_000, 5_000 * 2 ** attempt));
    } else if (res.status >= 500) {
      await sleep(500 * (attempt + 1));
    } else {
      throw new PortalError(502, `portal returned ${res.status}`);
    }
  }
  throw new PortalError(503, `portal kept failing (last status ${lastStatus})`);
}

export async function fetchReadings(args: { meterId: string; from?: string; to?: string }): Promise<RawReading[]> {
  const qs = new URLSearchParams();
  if (args.from) qs.set("from", args.from);
  if (args.to) qs.set("to", args.to);
  const path = `/portal/meters/${encodeURIComponent(args.meterId)}/energy?${qs}`;
  return (await get(path, z.object({ data: z.array(RawReading) }))).data;
}

export async function fetchTransformers(): Promise<Transformer[]> {
  const Page = z.object({ data: z.array(Transformer), total: z.number() });
  const all: Transformer[] = [];
  for (let page = 1; ; page++) {
    const res = await get(`/portal/dts?page=${page}`, Page);
    all.push(...res.data);
    if (all.length >= res.total || res.data.length === 0) return all;
  }
}

// Bulk export: every meter with hierarchy and location in one signed request.
export async function fetchExport(): Promise<ExportMeter[]> {
  const Keys = z.object({ data: z.object({ signingSecret: z.string() }) });
  for (let attempt = 0; attempt < 2; attempt++) {
    if (!signingSecret) signingSecret = (await get("/portal/keys", Keys)).data.signingSecret;
    const query = "page=1";
    const timestamp = String(Math.floor(Date.now() / 1000));
    const message = ["GET", "/portal/export", query, timestamp].join("\n");
    const signature = createHmac("sha256", signingSecret).update(message).digest("hex");
    try {
      const res = await get(`/portal/export?${query}`, z.object({ data: z.array(ExportMeter) }), {
        "x-timestamp": timestamp,
        "x-signature": signature,
      });
      return res.data;
    } catch (err) {
      if (err instanceof PortalError && err.status === 401 && attempt === 0) {
        signingSecret = ""; // secret may have rotated, refetch once
        continue;
      }
      throw err;
    }
  }
  throw new PortalError(502, "export signature rejected");
}
